crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -803,22 +814,77 @@ where
     }
 
     fn visit_pattern(&mut self, pattern: &'ast ast::Pattern) {
-        if let ast::Pattern::MatchAs(ast::PatternMatchAs {
-            name: Some(name), ..
-        })
-        | ast::Pattern::MatchStar(ast::PatternMatchStar {
+        // The definition visitor will recurse into the pattern so avoid walking it here.
+        let mut definition_visitor = MatchPatternDefinitionVisitor::new(self, pattern);
+        definition_visitor.visit_pattern(pattern);
+    }
+}
+
+/// A visitor that adds symbols and definitions for the identifiers in a match pattern.
+struct MatchPatternDefinitionVisitor<'a, 'db> {
+    /// The semantic index builder in which to add the symbols and definitions.
+    builder: &'a mut SemanticIndexBuilder<'db>,
+    /// The index of the current node in the pattern.
+    index: u32,
+    /// The pattern being visited. This pattern is the outermost pattern that is being visited
+    /// and is required to add the definitions.
+    pattern: &'a ast::Pattern,
+}
```

Comment:
The main reason for me to keep the visitor separate was because of the state variables which includes both the `index` and `pattern`. Here, the `pattern` is the outermost pattern being visited and is required to store because of the recursive nature of the visitor pattern. One advantage of using a separate visitor is that the state is reset as soon as the work is done. 

> The need to explicitly delegate `visit_expr` (and potentially other things in future?) seems like a downside of the separate-visitor approach. There are advantages to maintaining the invariant that everything will get visited by the same visitor.

I didn't realize that this delegation would be required until a bit later when I noticed some tests failing. That said, I think it makes sense to maintain this invariant to avoid any future failure like the one I mentioned. We could create a separate `State` struct which maintains the necessary variables.

Reply:
I'll make this change as a follow-up.

Reply:
https://github.com/astral-sh/ruff/pull/13209